File: ui/top_bar/sound.py
from gi.repository import Gtk


class Sound(Gtk.Box):
    def __init__(self, app, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.app = app
        self.app.pub_sub.subscribe(self)

        self.set_orientation(Gtk.Orientation.HORIZONTAL)
        self.set_spacing(5)
        self.set_css_classes(["widget", "sound", "padded"])
        self.set_name("Sound")

        self.image = Gtk.Image.new_from_icon_name("dialog-question")
        self.append(self.image)

    def on_volume(self, event):
        volume = event.volume
        muted = event.muted
        icon_name = None
        if volume == 0 or muted:
            icon_name = "audio-volume-muted-symbolic"
        elif volume >= 1 and volume < 34:
            icon_name = "audio-volume-low-symbolic"
        elif volume >= 34 and volume < 67:
            icon_name = "audio-volume-medium-symbolic"
        elif volume >= 67 and volume < 95:
            icon_name = "audio-volume-high-symbolic"
        else:
            icon_name = "audio-volume-overamplified-symbolic"
        self.image.set_from_icon_name(icon_name)

File: ui/src/top_bar/sound.c
#include "ui/include/top_bar/sound.h"
#include "ui/include/builder.h"

typedef struct {
  GtkWidget *image;
} data_t;
#define DATA_KEY "data"

GtkWidget *sound_init() {
  GtkWidget *self = top_bar_get_widget("SOUND");
  GtkWidget *image = top_bar_get_widget("SOUND_IMAGE");

  data_t *data = malloc(sizeof(data_t));
  data->image = image;
  g_object_set_data_full(G_OBJECT(self), DATA_KEY, data, free);

  return self;
}

void sound_refresh(GtkWidget *self, uint32_t volume, bool muted) {
  data_t *data = g_object_get_data(G_OBJECT(self), DATA_KEY);

  const char *icon_name = NULL;
  if (volume == 0 || muted) {
    icon_name = "audio-volume-muted-symbolic";
  } else if (volume >= 1 && volume < 34) {
    icon_name = "audio-volume-low-symbolic";
  } else if (volume >= 34 && volume < 67) {
    icon_name = "audio-volume-medium-symbolic";
  } else if (volume >= 67 && volume < 95) {
    icon_name = "audio-volume-high-symbolic";
  } else {
    icon_name = "audio-volume-overamplified-symbolic";
  }
  gtk_image_set_from_icon_name(GTK_IMAGE(data->image), icon_name);
}
